persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -1738,6 +1744,140 @@ repsertManySql ent n = putManySql' conflictColumns fields ent n
+  CopyField :: EntityField record typ -> HandleUpdateCollision record
+  -- | Only copy the field if it is not equal to the provided value.
+  CopyUnlessEq :: PersistField typ => EntityField record typ -> typ -> HandleUpdateCollision record
+
+-- | Postgres specific 'upsertWhere'. This method does the following:
+-- It will insert a record if no matching unique key exists.
+-- If a unique key exists, it will update the relevant field with a user-supplied value, however,
+-- it will only do this update on a user-supplied condition.
+-- For example, here's how this method could be called like such:
+--
+-- upsertWhere record [recordField =. newValue] [recordField /= newValue]
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+-- 
+-- @since 2.12.0.1
+upsertWhere
+  :: ( backend ~ PersistEntityBackend record
+     , PersistEntity record
+     , PersistEntityBackend record ~ SqlBackend
+     , MonadIO m
+     , PersistStore backend
+     , BackendCompatible SqlBackend backend
+     )
+  => record
+  -> SqlBackend
+  -> [Update record]
+  -> [Filter record]
+  -> ReaderT backend m ()
+upsertWhere record conn updates filts =
+  upsertManyWhere [record] conn [] updates filts
+
+-- | Postgres specific 'upsertManyWhere'. This method does the following:
+-- It will insert a record if no matching unique key exists.
+-- If a unique key exists, it will update the relevant field with a user-supplied value, however,
+-- it will only do this update on a user-supplied condition.
+-- For example, here's how this method could be called like such:
+--
+-- upsertManyWhere [record] [recordField =. newValue] [recordField /= newValue]
+--
+-- Called thusly, this method will insert a new record (if none exists) OR update a recordField with a new value
+-- assuming the condition in the last block is met.
+--
+-- -- @since 2.12.0.1
+upsertManyWhere ::
+  forall record backend m.
+  ( backend ~ PersistEntityBackend record,
+    BackendCompatible SqlBackend backend,
+    PersistEntityBackend record ~ SqlBackend,
+    PersistEntity record,
+    MonadIO m
+  ) =>
+  -- | A list of the records you want to insert, or update
+  [record] ->
+  SqlBackend ->
+  -- | A list of the fields you want to copy over.
+  [HandleUpdateCollision record] ->
+  -- | A list of the updates to apply that aren't dependent on the record being inserted.
+  [Update record] ->
+  [Filter record] ->
+  ReaderT backend m ()
+upsertManyWhere [] _ _ _ _ = return ()
+upsertManyWhere records conn fieldValues updates conditions =
+  uncurry rawExecute $
+    mkBulkUpsertQuery records conn fieldValues updates conditions
+
+-- | This creates the query for 'upsertManyWhere'. If you
+-- provide an empty list of updates to perform, then it will generate
+-- a dummy/no-op update using the first field of the record. This avoids
+-- duplicate key exceptions.
+mkBulkUpsertQuery
```

Comment:
this method was copied from `MySQL.hs` as well.  I just had to change some of the type constraints.